pkg/controller/migration/namespace_migration.go

```diff
@@ -220,18 +220,22 @@ func (m *CoreNamespaceMigration) Run(ctx context.Context, log logr.Logger) error
 		return fmt.Errorf("failed deleting kube-system calico-kube-controllers: %s", err.Error())
 	}
 	log.V(1).Info("Deleted previous calico-kube-controllers deployment")
-	if err := m.waitForOperatorTyphaDeploymentReady(ctx); err != nil {
-		return fmt.Errorf("failed to wait for operator typha deployment to be ready: %s", err.Error())
-	}
-	log.V(1).Info("Operator Typha Deployment is ready")
 	if err := m.labelUnmigratedNodes(ctx); err != nil {
 		return fmt.Errorf("failed to label unmigrated nodes: %s", err.Error())
 	}
 	log.V(1).Info("All unmigrated nodes labeled")
-	if err := m.ensureKubeSysNodeDaemonSetHasNodeSelectorAndIsReady(ctx); err != nil {
+	if err := m.ensureKubeSysNodeDaemonSetHasNodeSelectorAndIsReady(ctx, log); err != nil {
 		return fmt.Errorf("the kube-system node DaemonSet is not ready with the updated nodeSelector: %s", err.Error())
 	}
 	log.V(1).Info("Node selector added to kube-system node DaemonSet")
+	if err := m.ensureTyphaRoom(ctx, log); err != nil {
+		return fmt.Errorf("unable to ensure room for enough typhas: %s", err.Error())
+	}
+	log.V(1).Info("Ensured room for Typha deployments")
+	if err := m.waitForOperatorTyphaDeploymentReady(ctx, log); err != nil {
```

Comment:
With the Operator Typha now being subject to the migration label, won't this always work (assuming no restarting/errors) because there will be no Operator Typha's running at this point.

Reply:
In practice, I think one will get scheduled anyway before the auto-scaler kicks in, but otherwise yeah I think you're right it's not a super useful check here. It would be better to do this as part of migrateEachNode.